src/model/operation/wrapoperation.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module engine/model/operation/wrapoperation
+ */
+
+import Operation from './operation';
+import UnwrapOperation from './unwrapoperation';
+import Position from '../position';
+import Range from '../range';
+import Element from '../element';
+import { _insert, _move } from './utils';
+
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+
+/**
+ * Operation to wrap a range of {@link module:engine/model/item~Item model items} with
+ * a {@link module:engine/model/element~Element model element}.
+ *
+ * @extends module:engine/model/operation/operation~Operation
+ */
+export default class WrapOperation extends Operation {
+	/**
+	 * Creates a wrap operation.
+	 *
+	 * @param {module:engine/model/position~Position} position Position before
+	 * the first {@link module:engine/model/item~Item model item} to wrap.
+	 * @param {Number} howMany Offset size of wrapped range. Wrapped range will start at `position.offset` and end at
+	 * `position.offset + howMany`.
+	 * @param {module:engine/model/element~Element|module:engine/model/position~Position} elementOrPosition Element to
+	 * wrap with or position in graveyard before the element which should be used as a wrapper.
+	 * @param {Number|null} baseVersion Document {@link module:engine/model/document~Document#version} on which operation
+	 * can be applied or `null` if the operation operates on detached (non-document) tree.
+	 */
+	constructor( position, howMany, elementOrPosition, baseVersion ) {
+		super( baseVersion );
+
+		/**
+		 * Position before the first {@link module:engine/model/item~Item model item} to wrap.
+		 *
+		 * @member {module:engine/model/position~Position} module:engine/model/operation/wrapoperation~WrapOperation#position
+		 */
+		this.position = Position.createFromPosition( position );
+		this.position.stickiness = 'toNext';
+		// maybe change to a range
+
+		/**
+		 * Offset size of wrapped range.
+		 *
+		 * @member {Number} module:engine/model/operation/wrapoperation~WrapOperation#howMany
+		 */
+		this.howMany = howMany;
+
+		/**
+		 * Element to wrap with.
+		 *
+		 * @member {module:engine/model/element~Element} module:engine/model/operation/wrapoperation~WrapOperation#element
+		 */
+		this.element = elementOrPosition instanceof Element ? elementOrPosition : null;
+
+		this.graveyardPosition = elementOrPosition instanceof Element ? null : Position.createFromPosition( elementOrPosition );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	get type() {
+		return 'wrap';
+	}
```

Comment:
Consider if we still need types of opperations.

Reply:
I think we do, especially if we will join `RemoveOperation` and `ReinsertOperation` to `MoveOperation`.